### PR TITLE
[libc++] Add an ABI setting to harden unique_ptr<T[]>::operator[]

### PR DESCRIPTION
**Pull request by ldionne**
This allows catching OOB accesses inside `unique_ptr<T[]>` when the size
of the allocation is known. The size of the allocation can be known when
the unique_ptr has been created with make_unique & friends or when the
type necessitates an array cookie before the allocation.

### CODE REVIEW

**Comment by ldionne** [comment]
Note that this patch is kind of a WIP. Another approach could be to query the compiler or the system allocator for the size of the allocation. I actually think the compiler knows the size of the allocation cause it encodes it at the front of the allocation when calling `operator new[]`, but I have to refresh my memory on this.

**Comment by efriedma-quic** [comment]
> I actually think the compiler knows the size of the allocation cause it encodes it at the front of the allocation

Only when the element type has a non-trivial destructor.

File: libcxx/include/__memory/unique_ptr.h [hunk 1]
```diff
@@ -434,6 +462,10 @@ class _LIBCPP_UNIQUE_PTR_TRIVIAL_ABI _LIBCPP_TEMPLATE_VIS unique_ptr<_Tp[], _Dp>
   }
 
   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX23 __add_lvalue_reference_t<_Tp> operator[](size_t __i) const {
+#ifdef _LIBCPP_ABI_BOUNDED_UNIQUE_PTR
+    _LIBCPP_ASSERT_VALID_ELEMENT_ACCESS(
+        __size_ == 0 || __i < __size_, "unique_ptr<T[]>::operator[](index): index out of range");
```
Comment (haoNoQ) [inline libcxx/include/__memory/unique_ptr.h:467]: If you use `SIZE_MAX` as "unspecified size", will it eliminate the branch? 🤔

**Comment by EricWF** [comment]
A couple of things I would like us to investigate:

1. Can we stash the size before the the array we allocate?
2. What effect does this have on unique_ptr being passed in registers?

**Comment by jkorous-apple** [comment]
> > I actually think the compiler knows the size of the allocation cause it encodes it at the front of the allocation
> 
> Only when the element type has a non-trivial destructor.

That sounds like for such types we could use it for runtime check in `operator[]` without changing the layout. Is that right?
Do you see any catch?

**Comment by efriedma-quic** [comment]
https://itanium-cxx-abi.github.io/cxx-abi/abi.html#array-cookies describes array cookies.  If the cookie is present, I can't think of any reason it would be a problem for libc++ to read it.

If the cookie isn't present, though, we can't access it, and forcing the compiler to generate a cookie where it normally wouldn't generate one is ABI break.

**Comment by ldionne** [comment]
I updated the patch with a WIP attempt to use the array cookie when one is present. There are still things to fix and comments to address (e.g. maybe using `SIZE_MAX` is better to eliminate a branch), but I wanted to get this draft out there for comments. I think it's a pretty clever improvement to `unique_ptr`.

**Comment by ldionne** [comment]
@vitalybuka I just stumbled upon `-fno-sanitize-address-poison-custom-array-cookie` and learned that ASAN sometimes poisons array cookies when a custom `operator new` is in use. I'm not certain I understand why poisoning the array cookie is legal even when we have a custom `operator new[]`, but it seems like poisoning array cookies would interact badly with this patch. Do you have thoughts?

File: libcxx/include/__memory/array_cookie.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+// Under the Itanium C++ ABI [1], we know that an array cookie is available unless `T` is trivially
+// destructible and the call to `operator delete[]` is not a sized operator delete. Under ABIs other
+// than the Itanium ABI, we assume there are no array cookies.
+//
+// [1]: https://itanium-cxx-abi.github.io/cxx-abi/abi.html#array-cookies
+#ifdef _LIBCPP_ABI_ITANIUM
+// TODO: Use a builtin instead
+// TODO: We should factor in the choice of the usual deallocation function in this determination.
+template <class _Tp>
+struct __has_array_cookie : _Not<is_trivially_destructible<_Tp> > {};
```
Comment (ldionne) [inline libcxx/include/__memory/array_cookie.h:39]: @efriedma-quic What do you think about this? It is a bit more conservative than we could be, but I don't really want to start trying to implement the resolution of `operator delete[]` in the library. It looks like a compiler builtin should be fairly easy to provide for both getting the array cookie and telling us whether an array cookie is available.

I'm curious to know whether you think this is a reasonable first implementation.

Comment (efriedma-quic) [reply libcxx/include/__memory/array_cookie.h:39]: This seems reasonable for now.

**Comment by vitalybuka** [comment]
> @vitalybuka I just stumbled upon `-fno-sanitize-address-poison-custom-array-cookie` and learned that ASAN sometimes poisons array cookies when a custom `operator new` is in use. I'm not certain I understand why poisoning the array cookie is legal even when we have a custom `operator new[]`, but it seems like poisoning array cookies would interact badly with this patch. Do you have thoughts?

I see this feature for the first time :)

Looks like @filcab added the feature. Before that custom new was omitted, then
https://reviews.llvm.org/D41301 made it more aggressive
https://reviews.llvm.org/D52615 hidden the change above by the flag

And  the doc string was gone with https://reviews.llvm.org/D133349

I believe this is off by default "use at your own risk" feature.

I guess it should be easy to support this case by `__attribute__((no_sanitize("address")))` on __get_array_cookie

**Comment by ldionne** [comment]
> LLVM Buildbot has detected a new failure on builder `sanitizer-x86_64-linux-android` running on `sanitizer-buildbot-android` while building `libcxx` at step 2 "annotate".
> [...]

I looked into it and I don't think this is caused by the patch, since the test doesn't use `unique_ptr`.

**Comment by nico** [comment]
This makes the attached program crash (…in most runs, not 100% of the time):

[unique.cc.zip](https://github.com/user-attachments/files/17196807/unique.cc.zip)

```
 % out/gn/bin/clang++ unique.cc -isysroot $(xcrun -show-sdk-path) -std=c++20 -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_EXTENSIVE     
% ./a.out
zsh: trace trap  ./a.out
```

lldb shows that this indeed this check:

```
    frame #0: 0x00000001000029e4 a.out`std::__1::unique_ptr<std::__1::vector<int, std::__1::allocator<int>> [], AlignedDeleter<std::__1::vector<int, std::__1::allocator<int>>>>::operator[][abi:se200000](unsigned long) const [inlined] __clang_trap_msg$libc++$/Users/thakis/src/llvm-project/out/gn/bin/../include/c++/v1/__memory/unique_ptr.h:566: assertion __checker_.__in_bounds(std::__to_address(__ptr_), __i) failed: unique_ptr<T[]>::operator[](index): index out of range
```

The program uses a unique_ptr with a custom deleter to refer to raw malloc-ed memory, "to allow having uninitialized entries inside it".

It has this typedef:

```
template <typename T>
using AlignedUniquePtr =
    std::unique_ptr<T, AlignedDeleter<typename std::remove_extent<T>::type>>;
```

It uses this typedef with an array type:

```
  // Underlying storage. It's raw malloc-ed rather than being a unique_ptr<T[]>
  // to allow having uninitialized entries inside it.
  AlignedUniquePtr<T[]> entries_;
```

It creates objects of this unique_ptr like so:

```
template <typename T>
AlignedUniquePtr<T> AlignedAllocTyped(size_t n_membs) {
  using TU = typename std::remove_extent<T>::type;
  return AlignedUniquePtr<T>(
      static_cast<TU*>(AlignedAlloc(alignof(TU), sizeof(TU) * n_membs)));
}
```

`AlignedAlloc()` is basically malloc – in particular, nothing calls `new[]`, so it doesn't write an array cookie.

The custom deleter looks like so (`AlignedFree()` is just `free()`):

```
template <typename T>
struct AlignedDeleter {
  inline void operator()(T* ptr) const { AlignedFree(ptr); }
};
```

The class containing `entries_` calls dtors as needed, and the deleter really only needs to free memory. `delete []` is never called.

Is this a valid use of unique_ptr<T[]>? If so, should this hardening only be used for unique_ptrs that don't have a custom deleter?

**Comment by nico** [comment]
Looking through `20.3.1.4 unique_ptr for array objects with a runtime length [unique.ptr.runtime]`, it says "— The default deleter will call delete[]." which implies that looking for the array cookie on itanium is with the default deleter, but other than that it doesn't say anything. So I think the code in the example is probably correct, and this diagnostic isn't – with a custom deleter, it can introduce an out-of-bounds read for valid code.

**Comment by ldionne** [comment]
I think you're entirely correct. I'm working on a patch.

**Comment by ldionne** [comment]
In fact, the memory held by the `unique_ptr` would not even have to be allocated. It could point to a stack variable and the deleter could be a no-op. In that sense, `unique_ptr` with a custom deleter is a bit closer to a `scope_guard`. I think I had that in mind when I made the first implementation a few months, but when I picked it back up recently I remember thinking "oh well it works for custom deleters as well" -- but that was a mistake.

**Comment by alexfh** [comment]
What about a case when a `std::unique_ptr<T[]>` with a default deleter temporarily points to a memory allocated with something but `new T[]`? If the code ensures the pointer is always reset before `std::unique_ptr` is destroyed, and the default deleter is never used. Something like in this example: https://gcc.godbolt.org/z/4fczP6e1q

If we skip the question why would one do that at all (I'm about to ask the author(s)), it seems like this is a valid  use of the `std::unique_ptr<>` according to the C++20 standard. Or am I missing something?

File: libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.observers/assert.subscript.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  });
+}
+
+template <std::size_t Size>
+struct NoCookie {
+  char padding[Size];
+};
+
+template <std::size_t Size>
+struct WithCookie {
```
Comment (var-const) [inline libcxx/test/std/utilities/smartptr/unique.ptr/unique.ptr.class/unique.ptr.observers/assert.subscript.pass.cpp:146]: Optional: can we `static_assert` these properties? (it's already a libc++-specific test, so I think it's acceptable to access internal types)

File: libcxx/include/__memory/unique_ptr.h [hunk 4]
```diff
@@ -292,6 +298,91 @@ class _LIBCPP_UNIQUE_PTR_TRIVIAL_ABI _LIBCPP_TEMPLATE_VIS unique_ptr {
   }
 };
 
+// Bounds checking in unique_ptr<T[]>
```
Comment (var-const) [inline libcxx/include/__memory/unique_ptr.h:301]: Can you please also update the hardening docs and perhaps add a release note?

File: libcxx/include/__memory/unique_ptr.h [hunk 5]
```diff
@@ -411,11 +518,13 @@ class _LIBCPP_UNIQUE_PTR_TRIVIAL_ABI _LIBCPP_TEMPLATE_VIS unique_ptr<_Tp[], _Dp>
...
   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX23 unique_ptr(unique_ptr&& __u) _NOEXCEPT
       : __ptr_(__u.release()),
-        __deleter_(std::forward<deleter_type>(__u.get_deleter())) {}
+        __deleter_(std::forward<deleter_type>(__u.get_deleter())),
+        __checker_(std::move(__u.__checker_)) {}
 
   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX23 unique_ptr& operator=(unique_ptr&& __u) _NOEXCEPT {
     reset(__u.release());
     __deleter_ = std::forward<deleter_type>(__u.get_deleter());
+    __checker_ = std::move(std::move(__u.__checker_));
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/__memory/unique_ptr.h:527]: Yes, looks like a typo 😱

File: libcxx/include/__memory/unique_ptr.h [hunk 6]
```diff
@@ -292,6 +298,91 @@ class _LIBCPP_UNIQUE_PTR_TRIVIAL_ABI _LIBCPP_TEMPLATE_VIS unique_ptr {
...
+};
+
+// (2) Implementation where we store the size in the class whenever we have it.
+//
+// Semantically, we'd need to store the size as an optional<size_t>. However, since that
+// is really heavy weight, we instead store a size_t and use SIZE_MAX as a magic value
+// meaning that we don't know the size.
+struct __unique_ptr_array_bounds_stored {
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR __unique_ptr_array_bounds_stored() : __size_(SIZE_MAX) {}
+  _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR explicit __unique_ptr_array_bounds_stored(size_t __size) : __size_(__size) {}
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/__memory/unique_ptr.h:365]: `SIZE_MAX` is supposed to be defined in `<cstdint>`, can you check why your platform doesn't provide it?

**Comment by ldionne** [comment]
@var-const Thanks for the comments! See https://github.com/llvm/llvm-project/pull/111704.

**Comment by alexfh** [comment]
Thanks for the detailed analysis. I totally agree with the conclusions here. The value of this hardening is obvious, while the ability to (mis-)use unique_ptr this way is questionable on its own. Luckily (though no surprise ;) this was an isolated case and the code has already been fixed.

> To be unambiguous, we should perhaps add something to the various `std::unique_ptr` constructors that says "the pointer used to initialized the `unique_ptr` must be such that calling the deleter on it disposes of the pointer in a valid way", but we'd have to be careful not to step on the toes of `unique_ptr`'s ability to hold incomplete types.

Making this part of the standard unambiguous would definitely be welcome.

> Frankly, I'm not certain what to do with this. I feel like it would be really unfortunate to drop this hardening check on a slightly ambiguous technicality since it can deliver a ton of value in the real world. I think I'll email the reflector to see what the sentiment is like over there.

I almost feel sorry for bringing this up ;) But hopefully, this will result in improving the standard rather than creating an obstacle for library hardenings.

**Pull request merged**